internal/api/auth_hooks.go

```diff
@@ -0,0 +1,149 @@
+package api
+
+import (
+	"encoding/json"
+	"errors"
+	"net/url"
+	"time"
+
+	"fmt"
+	"github.com/gofrs/uuid"
+	"github.com/supabase/gotrue/internal/conf"
+	"github.com/supabase/gotrue/internal/storage"
+	"strings"
+)
+
+type HookType string
+
+const (
+	PostgresHook HookType = "postgres"
+	HTTPHook     HookType = "http"
+)
+
+type AuthHook struct {
+	*conf.ExtensibilityPointConfiguration
+	payload  []byte
+	hookType HookType
+	event    string
+	db       *storage.Connection
+}
+
+// Hook Events
+const (
+	MFAVerificationEvent = "auth.mfa_verfication"
+)
+
+const (
+	defaultTimeout = time.Second * 2
+)
+
+type HookErrorResponse struct {
+	ErrorMessage string `json:"error_message"`
+	ErrorCode    string `json:"error_code"`
+	RetryAfter   bool   `json:"retry_after"`
+}
+
+type MFAVerificationHookResponse struct {
+	Decision string `json:"decision"`
+}
+
+func parseErrorResponse(response []byte) (*HookErrorResponse, error) {
+	var errResp HookErrorResponse
+	err := json.Unmarshal(response, &errResp)
+	if err != nil {
+		return nil, err
+	}
+	if errResp.ErrorMessage != "" {
+		return &errResp, nil
+	}
+	return nil, err
+}
+
+func parseMFAVerificationResponse(response []byte) (*MFAVerificationHookResponse, error) {
+	var MFAVerificationResponse MFAVerificationHookResponse
+	err := json.Unmarshal(response, &MFAVerificationResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MFAVerificationResponse, err
+}
+
+// Functions for encoding and decoding payload
+func CreateMFAVerificationHookInput(user_id uuid.UUID, factor_id uuid.UUID, valid bool) ([]byte, error) {
+	payload := struct {
+		UserID   uuid.UUID `json:"user_id"`
+		FactorID uuid.UUID `json:"factor_id"`
+		Valid    bool      `json:"valid"`
+	}{
+		UserID:   user_id,
+		FactorID: factor_id,
+		Valid:    valid,
+	}
+	data, err := json.Marshal(&payload)
+	if err != nil {
+		return nil, err
+	}
+	return data, nil
```

Comment:
One thing about this is, this will probably never fail. We know what the inputs are and that the struct is marshallable. So maybe just `panic` instead and not return an error?

Reply:
Makes sense, thanks